tt_metal/impl/buffers/buffer.cpp

```diff
@@ -24,6 +24,12 @@ namespace tt {
 
 namespace tt_metal {
 
+std::ostream& operator<<(std::ostream& os, const ShardSpec& spec) {
+    os << "ShardSpec(grid=?"
+       << ", shape={" << spec.shape[0] << ", " << spec.shape[1] << "})";
+    return os;
```

Comment:
@tt-aho, can you share how this works? Would this rely on reflect lib?

I tried using it here and it failed with segfault.

Reply:
Yes, it does depend on the reflect library.
Was my mistake, the segfault issue is that our reflect lib will check if `operator<<` is defined for an object and use it, otherwise it will construct the string from our defined attributes (what currently happens when we use fmt::format), so there is a circular dependency when we add fmt::format inside operator<<. Was hoping to avoid manually defining a string here since we could already have one generated for us but doesn't look like it'll work out.

Could you format your string here in operator<< to output the string like how it's currently generated when using fmt::format? The current string doesn't print out all attributes and this will affect existing usage of printing ShardSpec.

Reply:
Actually, could you try this instead of using fmt::format? This should generate the same string as format without the circular dependency.
```
tt::stl::reflection::operator<<(os, spec);
return os;
```


The following function is what's called when we use fmt::format on our object. The line `ss << object` ends up invoking our defined `operator<<` instead of `tt::stl::reflection::operator<<`, resulting in the circular dependency, though maybe this is what was intended.

```
template <typename T>
struct fmt::formatter<T, char, std::enable_if_t<tt::stl::reflection::detail::supports_conversion_to_string_v<T>>> {
    constexpr auto parse(format_parse_context& ctx) -> format_parse_context::iterator { return ctx.end(); }

    auto format(const T& object, format_context& ctx) const -> format_context::iterator {
        using tt::stl::reflection::operator<<;
        std::stringstream ss;
        ss << object;
        return fmt::format_to(ctx.out(), "{}", ss.str());
    }
};
```

Reply:
Thank you!